exercises/dotnet/encryption-context-start/src/App.cs

```diff
@@ -0,0 +1,149 @@
+﻿using System.Text;
+using Amazon.DynamoDBv2;
+using Amazon.KeyManagementService;
+using Amazon.S3;
+using AWS.EncryptionSDK.Core;
+
+namespace DocumentBucket
+{
+
+    public class App
+    {
+        private static volatile bool _continue = true;
+
+        public static async Task Main()
+        {
+            string tableName = Config.TableName;
+            string bucketName = Config.BucketName;
+            string region = Config.Region;
+
+            AmazonDynamoDBConfig amazonDynamoDBConfig = new();
+            amazonDynamoDBConfig.RegionEndpoint = Amazon.RegionEndpoint.GetBySystemName(region);
+            AmazonDynamoDBClient amazonDynamoDBClient = new(amazonDynamoDBConfig);
+
+            AmazonS3Config amazonS3Config = new();
+            amazonS3Config.RegionEndpoint = Amazon.RegionEndpoint.GetBySystemName(region);
+            AmazonS3Client amazonS3Client = new(amazonS3Config);
+
+            var materialProviders = AwsCryptographicMaterialProvidersFactory.CreateDefaultAwsCryptographicMaterialProviders();
+            var keyring = materialProviders.CreateAwsKmsMultiKeyring(new CreateAwsKmsMultiKeyringInput
+            {
+                Generator = Config.FaytheKmsKeyId,
+                KmsKeyIds = new List<string>() {Config.WalterKmsKeyId}
+            });
+
+            Api api = new(amazonDynamoDBClient, tableName, amazonS3Client, bucketName, keyring);
+
+            Console.CancelKeyPress += new ConsoleCancelEventHandler(CancelHandler);
+            while (_continue)
+            {
+                Console.WriteLine("Enter a number to chose a menu option:");
+                Console.WriteLine("1. List");
+                Console.WriteLine("2. Store");
+                Console.WriteLine("3. Retrieve");
+                try
+                {
+                    switch (Console.ReadLine())
+                    {
+                        case "1":
+                            var items = await api.List();
+                            await ListToConsole(api, items);
+                            break;
+                        case "2":
+                            Console.WriteLine("Enter data to store:");
+                            var data = Console.ReadLine();
+                            if (data is null || data == "")
+                            {
+                                Console.WriteLine("You must enter a valid input!");
+                                break;
+                            }
+                            Console.WriteLine($"Storing: {data}");
+                            var context = ReadEncryptionContext();
+                            if (context.Count > 0)
+                            {
+                                Console.WriteLine("Adding Encryption Context:");
+                                Console.WriteLine(string.Join(Environment.NewLine, context));
+                            }
+                            await api.Store(Encoding.ASCII.GetBytes(data), context);
```

Comment:
If we only intend to accept valid ASCII as data and encryption context keys/values, it's better to validate that immediately upon reading it from stdin. Otherwise, [invalid ASCII characters will silently be replaced by `?`](https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.ascii?view=net-6.0#remarks).

Reply:
Good call - I moved to using UTF8 for the data for more robust character support and introduced validation for ASCII for encryption context (which must be ASCII since it gets added to the request header).